L1Trigger/L1THGCal/plugins/be_algorithms/HGCalTriggerClusterIdentificationBDT.cc

```diff
@@ -0,0 +1,168 @@
+    int category(float pt, float eta) const;
+
+
+};
+
+DEFINE_EDM_PLUGIN(HGCalTriggerClusterIdentificationFactory,
+    HGCalTriggerClusterIdentificationBDT,
+    "HGCalTriggerClusterIdentificationBDT" );
+
+
+HGCalTriggerClusterIdentificationBDT::
+HGCalTriggerClusterIdentificationBDT():HGCalTriggerClusterIdentificationBase()
+{
+}
+
+void
+HGCalTriggerClusterIdentificationBDT::
+initialize(const edm::ParameterSet& conf)
+{
+  categories_.clear();
+  bdts_.clear();
+  working_points_.clear();
+  input_variables_.clear();
+  input_variables_ = conf.getParameter< std::vector<std::string> >("Inputs");
+  std::vector<std::string> bdt_files = conf.getParameter< std::vector<std::string> >("Weights");
+  std::vector<double> categories_etamin = conf.getParameter<std::vector<double>>("CategoriesEtaMin");
+  std::vector<double> categories_etamax = conf.getParameter<std::vector<double>>("CategoriesEtaMax");
+  std::vector<double> categories_ptmin = conf.getParameter<std::vector<double>>("CategoriesPtMin");
+  std::vector<double> categories_ptmax = conf.getParameter<std::vector<double>>("CategoriesPtMax");
+  working_points_ = conf.getParameter<std::vector<double>>("WorkingPoints");
+  for(unsigned cat=0; cat<categories_etamin.size(); cat++)
+  {
+    categories_.emplace_back(
+        categories_ptmin[cat],
+        categories_ptmax[cat],
+        categories_etamin[cat],
+        categories_etamax[cat]);
+  }
+  std::vector<std::string> spectators = {};
+  for (const auto& file : bdt_files)
+  {
+    bdts_.emplace_back(new TMVAEvaluator());
+    bdts_.back()->initialize(
+        "!Color:Silent:!Error",
+        "BDT::bdt",
+        edm::FileInPath(file).fullPath(),
+        input_variables_,
+        spectators,
+        false, false);
+  }
+}
+
+float
+HGCalTriggerClusterIdentificationBDT::
+value(const l1t::HGCalMulticluster& cluster) const
+{
+  std::map<std::string, float> inputs;
+  for(const std::string& var : input_variables_)
+  {
+    inputs[var] = clusterVariable(var, cluster);
+  }
+  float pt = cluster.pt();
+  float eta = cluster.eta();
+  int cat = category(pt, eta);
+  return (cat!=-1 ? bdts_.at(cat)->evaluate(inputs) : -999.);
+}
+
+
+bool
+HGCalTriggerClusterIdentificationBDT::
+decision(const l1t::HGCalMulticluster& cluster) const
+{
+  float bdt_output = value(cluster);
+  float pt = cluster.pt();
+  float eta = cluster.eta();
+  int cat = category(pt, eta);
+  return (cat!=-1 ? bdt_output>working_points_.at(cat) : true);
+}
+
+int 
+HGCalTriggerClusterIdentificationBDT::
+category(float pt, float eta) const
+{
+  for(unsigned cat=0; cat<categories_.size(); cat++)
+  {
+    if(categories_[cat].contains(pt, eta)) return static_cast<int>(cat);
+  }
+  return -1;
+}
+
+
+
+
+float
+HGCalTriggerClusterIdentificationBDT::
+clusterVariable(const std::string& variable, const l1t::HGCalMulticluster& cluster) const
+{
+  if(variable=="cl3d_showerlength") return cluster.showerLength();
```

Comment:
think about defining enums for these known quantities to limit repeated string comparisons